custom_model_runner/datarobot_drum/drum/gpu_predictors/nemo_predictor.py

```diff
@@ -0,0 +1,250 @@
+from datarobot_drum.drum.common import SupportedPayloadFormats
+from datarobot_drum.drum.enum import LOGGER_NAME_PREFIX, PayloadFormat, StructuredDtoKeys
+from datarobot_drum.drum.exceptions import DrumCommonException
+from datarobot_drum.drum.language_predictors.base_language_predictor import BaseLanguagePredictor
+from datarobot_drum.resource.drum_server_utils import DrumServerProcess, wait_for_server
+
+RUNNING_LANG_MSG = "Running environment: NeMo Inferencing Microservice."
+DEFAULT_MODEL_NAME = "generic_llm"
+
+
+class ChatRoles:
+    SYSTEM = "system"
+    ASSISTANT = "assistant"
+    USER = "user"
+
+
+class NemoPredictor(BaseLanguagePredictor):
+    def __init__(self):
+        super(NemoPredictor, self).__init__()
+        self.logger = logging.getLogger(LOGGER_NAME_PREFIX + "." + __name__)
+
+        # chat input fields
+        self.system_prompt_value = self.get_optional_parameter("system_prompt")
+        self.user_prompt_column = self.get_optional_parameter("prompt_column_name", "promptText")
+        self.assistant_response_column = self.get_optional_parameter(
+            "assistant_column_name", "assistant"
+        )
+
+        # completions configuration can be changed with Runtime parameters
+        self.startup_timeout_sec = self.get_optional_parameter("startup_timeout_sec", 60)
+        self.max_tokens = self.get_optional_parameter("max_tokens", 512)
+        self.use_chat_context = self.get_optional_parameter("chat_context", False)
+        self.num_choices_per_completion = self.get_optional_parameter("n", 1)
+        self.temperature = self.get_optional_parameter("temperature", 0.01)
+
+        # Nemo server configuration is set in the Drop-in environment
+        self.gpu_count = os.environ.get("GPU_COUNT")
+        if not self.gpu_count:
+            raise ValueError("Unexpected empty GPU count.")
+        self.health_port = os.environ.get("HEALTH_PORT", "9997")
+        self.openai_port = os.environ.get("OPENAI_PORT", "9999")
+        self.nemo_host = os.environ.get("NEMO_HOST", "http://localhost")
+        self.nemo_port = os.environ.get("NEMO_PORT", "9998")
+        self.model_name = os.environ.get("MODEL_NAME", DEFAULT_MODEL_NAME)
+        self.nemo_process = None
+        self.nim_client = None
+
+    def mlpiper_configure(self, params):
+        super(NemoPredictor, self).mlpiper_configure(params)
+
+        # start Nemo server and check health it's health
+        self.nemo_process = DrumServerProcess()
+        self.nemo_thread = Thread(target=self._run_nemo_server, args=(self.nemo_process,))
+        atexit.register(self._shutdown_nemo)
+        self.nemo_thread.start()
+        self._check_nemo_health()
+        self.nim_client = OpenAI(base_url=f"{self.nemo_host}:{self.openai_port}/v1", api_key="fake")
+
+    @staticmethod
+    def get_optional_parameter(key, default_value=None):
+        try:
+            return RuntimeParameters.get(key)
+        except ValueError:
+            return default_value
```

Comment:
We need to think about this more. Either we need another interface for this or we need to add support for Execution Environments to support **defining** runtime params, and then the custom model will be a combination of the runtime param defs from the selected environment version and also from any defined in the model itself.

Reply:
I'll create a ticket to track that work